### ISSUE TITLE
LRU cache is broken

### ISSUE THREAD

**Issue by prixone** [mattermost/platform#6120]
#### Summary
The [utils/lru.go](https://github.com/mattermost/platform/blob/master/utils/lru.go) cache is not working properly for changes that happen too often. 

#### Steps to reproduce
1. Open mattermost on chrome, preferable latest version using account A
2. Open mattermost on firefox, using account B
3. Wait both to see each other as online
4. Close chrome completely using the X button without terminating or touching the mattermost tab
5. Client A is now stuck with forever online status

#### Expected behavior
Client A would be cached/broadcast as offline

#### Observed behavior
Client A does not get cached/broadcast as offline

#### Possible fixes
If you look at the [status.go](https://github.com/mattermost/platform/blob/master/app/status.go#L208) you will notice the following line `AddStatusCache(status)`  which is responsible to update the cache for that user. However, the cache even thought it has been updated, it is some how reverted back to its original state.

If you further make a new function, for example:
```
func ForceUserStatusUpdate(status *model.Status) {
	statusCache.Remove(status.UserId)
	statusCache.Add(status.UserId, status)
}
```

Note that if you output the value of the model status, before using `AddStatusCache(status)`, you will get the correct data(in both scenarios, using the force update above and using the regular `AddStatusCache(status)`), and if you use the `GetStatusFromCache` right after the call to `AddStatusCache(status)` and again output it, u will also get the proper data updated, however it will be reverted back once the API calls for `GetUserStatusesByIds`.

What I found out, is that if I call `AddStatusCache(status)` then use the method `GetStatusFromCache` and then call `AddStatusCache(status)` again, it will in some cases work, its also not accurate.

So for the time being what I decided to do was to skip the cache for the GetUserStatusesByIds and read it directly from the database, until the issue with the LRU is known / fixed.

```
//GetUserStatusesByIds used by apiV4
func GetUserStatusesByIds(userIds []string) ([]*model.Status, *model.AppError) {
	var statusMap []*model.Status
	metrics := einterfaces.GetMetricsInterface()

	if result := <-Srv.Store.Status().GetByIds(missingUserIds); result.Err != nil {
		return nil, result.Err
	} else {
		statuses := result.Data.([]*model.Status)

		for _, s := range statuses {
			AddStatusCache(s)
		}

		statusMap = append(statusMap, statuses...)
	}

	// For the case where the user does not have a row in the Status table and cache
	for _, userId := range userIds {
		if _, ok := statusMap[userId]; !ok {
			statusMap[userId] = model.STATUS_OFFLINE
		}
	}

	return statusMap, nil
}
```

Another thing to note is the [web_hub.go](https://github.com/mattermost/platform/blob/master/app/web_hub.go#L255), the unregister method seems a bit ambiguous there and if you do not change it, it will not call the Offline in some situations where it should have, change as follow:

```
			case webCon := <-h.unregister:
				userId := webCon.UserId

				devicesCount := 0
				indexToDel := -1
				for i, webConCandidate := range h.connections {
					// we want to check for the UserId to count devices connected
					if userId == webConCandidate.UserId {
						devicesCount++

						// if the WebCon matches, we want to set it to be deleted
						if webConCandidate == webCon {
							indexToDel = i
						}

						// We already know there is more than 1 device, 
						// and we already have the webConCandiate to delete, so break here
						if devicesCount > 1 && indexToDel != -1 {
							break;
						}
					}
				}

				if indexToDel != -1 {
					// Delete the webcon we are unregistering
					h.connections[indexToDel] = h.connections[len(h.connections)-1]
					h.connections = h.connections[:len(h.connections)-1]
				}

				if devicesCount == 1 {
					go SetStatusOffline(userId, false)
				}
```
Unhappily I wasn't getting any answers on mattermost's Developers's channel, so I couldn't further look on how to fix it, thus posting it here for whoever takes over on the issue.

**Comment by fjarlq** [comment]
Regarding [the `web_hub.go` code](https://github.com/mattermost/platform/blob/master/app/web_hub.go#L255) you're wondering about, I think it is trying to avoid calling `SetStatusOffline` in the case where user A has two connections to the server but only one of them is being unregistered. Setting `found` to `true` means there is another connection by that user, so `SetStatusOffline` shouldn't be called. When `found` remains `false`, that means it didn't find another connection with the same `UserId`, so in that case it's correct to call `SetStatusOffline`.

**Comment by prixone** [comment]
@fjarlq that makes sense, unhappily it triggers true with single connections on client A. Since I only tested with a single client, so that behavior shouldn't happen from what you have explained above.

**Comment by jasonblais** [comment]
Thanks @prixone! Great issue description! 

I think your issue got unnoticed because of Easter holidays, sorry about it. I've been able to reproduce your issue though and have created a ticket: https://mattermost.atlassian.net/browse/PLT-6333

**Comment by prixone** [comment]
@jasonblais it's not a fix, its a workaround, where it will simple use the database calls instead of the cache, since the problem with the cache seems more complicated than my golang knowledge and I was not able to reach anyone to explain it to me, all I can tell is that its malfunctioning for updates that are very short in time from one to another. I've also updated the web_hub.go code above which seems to work for me but needs more testing with multiple devices to ensure its consistency.

**Comment by prixone** [comment]
I don't have a way to comment on the jira ticket so I will do it here.

@jasonblais 60 seconds is by itself long honestly should be within 15 second at most 30 seconds.

@lindalumitchell on your test did you run the server and clients all from the same machine? On the tests I did, I've run each from separated machines and the result was much longer than that.

Client A was on VM A with IP XYZ
Client B was on VM B with IP ABC
Mattermost Server was on VM C with IP EFG with nginx running on the same VM for proxying it

**Comment by lindalumitchell** [comment]
@prixone yes, I ran all on the same machine. I can re-test separately if we want to add more info / repro your results, but I agree with you that even 60+ seconds feels long. I believe the expectation was closer to 15 seconds.

**Comment by lindy65** [comment]
Hi everyone, I'm going to close this issue off here for now as there is a [Jira ticket](https://mattermost.atlassian.net/browse/PLT-6333) tracking a fix scheduled for v3.10.
